### PR TITLE
Use a Docker volume `bazel-output` for bazel-container output.

### PR DESCRIPTION
**Pull request by SanjayVas**
This is a behavioral change where the build output of the script is isolated from the host filesystem. The workspace directory is mounted as readonly and any convenience symlinks are left untouched. The script still bind mounts the Bazelisk cache and repository cache from the host filesystem.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362)

### CODE REVIEW

**Comment by SanjayVas** [comment]
This is intended to be merged into main after parent PR #361.

This should hopefully resolve any remaining issues w.r.t. running on MacOS, as I suspect the problem was that some of Bazel's filesystem operations are not properly implemented in the bind mount driver. See https://docs.docker.com/storage/volumes/ for more info on Docker volumes.

**Review by wangyaopw — COMMENTED** [review]
On my glinux machine, it fails with

```
$ tools/bazel-container test src/...
Running in Bazel container with output volume bazel-output: build src/...
FATAL: mkdir('/home/wangyaopw/.cache/bazel/_bazel_wangyaopw'): (error: 13): Permission denied
```
`sudo tools/bazel-container test src/...` works. Do you want to add `sudo` to the script? or do we manually add it when necessary?


On my Macbook, the `tools/bazel-container build src/...` passe. But the GCloudInProcessLifeOfAMeasurementIntegrationTest fails due to TIMEOUT now. There are logs like

```
Nov 13, 2021 6:46:58 AM org.wfanet.measurement.duchy.daemon.mill.MillBase$handleExceptions$2 invokeSuspend
SEVERE: YBGuy-fz_O0@worker2 liquidLegionsV2 Mill: TRANSIENT error
io.grpc.StatusException: FAILED_PRECONDITION: Cannot write timestamps in the future, found: 2021-11-13T06:46:58.389783+00:00
	at io.grpc.Status.asException(Status.java:549)
	at io.grpc.kotlin.ClientCalls$rpcImpl$1$1$1.onClose(ClientCalls.kt:289)
```
But I think this is not related to this PR. Since it fails both running with the script and running manually inside the container.

Timeout only happens on the MAC.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362#-:-MoN1P_Z4z0hYdAN8EvA:b-mde5a7)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @chipingyeh)

**Review by SanjayVas — COMMENTED** [review]
> sudo tools/bazel-container test src/... works. Do you want to add sudo to the script? or do we manually add it when necessary?

It shouldn't be needed at all. It's also surprising, as that path should only need to be created within the container. I'll do some more digging.

> io.grpc.StatusException: FAILED_PRECONDITION: Cannot write timestamps in the future, found: 2021-11-13T06:46:58.389783+00:00

Good catch. Yes, not the PR but might help debug what's going on. I'm guessing something to do with the Spanner emulator and commit timestamps.

> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362#-:-MoZNN8V0uU9EbWpFxdY:buvqidr)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @chipingyeh)

**Review by wangyaopw — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @chipingyeh)

**Review by SanjayVas — COMMENTED** [review]
> sudo tools/bazel-container test src/... works. Do you want to add sudo to the script? or do we manually add it when necessary?
> 
> It shouldn't be needed at all. It's also surprising, as that path should only need to be created within the container. I'll do some more digging.

Ah, I believe it's an issue when the Docker daemon is run as root. The script has to set the user for the container to be the current user so that any writes to the shared mounts have the right file ownership. The problem is that `/home` is owned by root and Docker inherits permissions/ownership from the parent folder when mounting. The solution is mounting inside of a globally-writable path such as `/tmp`, so I changed the script to do this.

I didn't hit the issue myself as I'm using rootless Docker.

> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362#-:-MoZOlJU9dyv1CYm2G2T:bpebcuo)* status: 0 of 1 files reviewed, all discussions resolved (waiting on @chipingyeh and @wangyaopw)

**Review by chipingyeh — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362#-Mo_GnTEDuXG55sCcWVw:-Mo_GnTEDuXG55sCcWVx:b-tz2f20) (no related file):*
Thanks for doing this! It now builds for me, although testing still fails on 1 integration test (GCloudInProcessLifeOfAMeasurementIntegrationTest). 

---

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362)* status: all files reviewed, 1 unresolved discussion (waiting on @chipingyeh and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362#-Mo_GnTEDuXG55sCcWVw:-Mo_HfZ46MHrLnlkpqf2:bcih1qh) (no related file):*
<details><summary><i>Previously, chipingyeh wrote…</i></summary><blockquote>

Thanks for doing this! It now builds for me, although testing still fails on 1 integration test (GCloudInProcessLifeOfAMeasurementIntegrationTest). 
</blockquote></details>

Np. I'll hopefully have access to a Mac on Wednesday to try to fix that part.

Note that the parent PR #361 needs your approval before this PR can be merged.

---

**Comment by chipingyeh** [comment]
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/362#-Mo_GnTEDuXG55sCcWVw:-Mo_IsD98H41mOx-bbez:b-896fix) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Np. I'll hopefully have access to a Mac on Wednesday to try to fix that part.

Note that the parent PR #361 needs your approval before this PR can be merged.
</blockquote></details>

Done.

---

**Comment by SanjayVas** [comment]
As a tip for those using this script on Linux (e.g. to build with an older glibc) who want to run any of the output on the host machine, you can use `docker volume inspect bazel-output` to find the Bazel output path on the host machine and run things from its execroot.

For example, the Mountpoint returned for me is `~/.local/share/docker/volumes/bazel-output/_data`. My execroot ends up at `~/.local/share/docker/volumes/bazel-output/_data/b2b94c5fcd8d59643147af90b9148cbf/execroot/wfa_measurement_system/bazel-out/k8-fastbuild/bin`, so I can `cd` there and run any executables I've build using the script.

**Comment by SanjayVas** [comment]
#378 changes this, and includes documentation for the correct steps.

**Pull request merged**
